man/watchsys.1
.TH watchsys 1

.SH NAME
watchsys \- script for basic system monitoring and prevention.

.SH SYNOPSIS
.B watchsys
[OPTION]

.SH DESCRIPTION
.PP
WatchSys is a set of bash shell scripts that run from a central daemon 
for basic monitoring of a linux server processes, cpu usage, 
network connections, ram usage, disk usage, file changes, etc...

.SH OPTIONS

.TP
\fB\-h\fR, \fB\-\-help\fR
display the script options
.TP
\fB\-d\fR, \fB\-\-start\fR
initialize a daemon to monitor the system
.TP
\fB\-s\fR, \fB\-\-stop\fR
stop the daemon
.TP
\fB\-t\fR, \fB\-\-status\fR
display the daemon status and pid if running

.SH SERVICES

.PP
On 
.B /etc/watchsys/proc.list
You can add a list of services to monitor. If one is down it 
will be restarted by executing the default system command,
which is auto-detected and can be: 

.TP
1. systemctl start <service_name>
.TP
2. service <service_name>
.TP
3. /etc/init.d/<service_name>

.PP
Also, if you are monitoring a custom process you can specify a custom
command to restart it.

Example content of this file:
 
# [Process Name] : [Service Name] : [Start Command]
.TP
apache : apache2 : default
.TP
hiawatha : hiawatha : -c /etc/hiawatha/hiawatha-custom.conf
.TP
named : bind9 : default

.SH SERVERS

.PP
On 
.B /etc/watchsys/server.list 
you can put a list of hostnames or ip's which 
are ping to see if they are alive.

Example of this file:

# [Server] : [Port]
.TP
my-domain.com : none
.TP
my-domain.com : 80
.TP
mail.my-domain.com : 143

.PP
As you see 
.B none 
is used when we want to monitor a server online
status, but not a service on a specific port.

.SH SETTINGS
You can modify the script settings by editing /etc/watchsys/watchsys.conf

.TP
.B EMAIL_TO = '<email>'
An email is sent to the following address when an important
system event is triggered, like critical cpu or ram usage,
service crashes, network status, file changes, etc...
(Default: root)

.TP
.B CUSTOM_EMAIL_COMMAND = '<command>'
If a custom e-mail command is given, it will be used to send all
e-mail notifications. This command will receive 3 parameters:
$email, $subject and $email_message_file. $email_message_file is
a path to a file that contains the message text. For example:
If you set the command to /usr/bin/mycommand it will be called
as follows: 

/usr/bin/mycommand $email $subject $email_message_file

Leave empty for default. (Default: "mail")

.TP
.B ALLOW_THREADING = <0|1>
Setting this value to 1 will launch a separate instance everytime
monitoring of a resource is needed which is faster than monitoring
in sequence. Set to 0 to disable. (Default: 1)

.PP
For other configuration options make sure to check
.B /etc/watchsys/watchsys.conf
which is documented with other options to control
monitor and e-mail intervals as values used when considering
a warning or critical status for a resource.

.SH FILES
.B /usr/share/doc/watchsys/LICENSE
.br
.B /usr/lib/watchsys/
.br
.B /usr/bin/watchsys
.br
.B /etc/watchsys/watchsys.conf
.br
.B /etc/watchsys/proc.list
.br
.B /etc/watchsys/server.list
.br
.B /etc/watchsys/dir.list

.SH AUTHOR
Jefferson González <jgmdev@gmail.com>

.SH LICENSE
GPLv3 <http://www.gnu.org/licenses/gpl-3.0.html>


ChangeLog
watchsys (0.1) stable; urgency=low
  
  * Initial version

 -- Jefferson González <jgmdev@gmail.com>  Fri, 11 Jan 2015 17:27:01 +0200


src/watchsys.sh
#!/bin/bash
########################################################################
# author: Jefferson González <jgmdev@gmail.com>                        #
# copyright: 2015 Jefferson González                                   #
#                                                                      #
# This file is part of WatchSys, which is under the GPLv3 License.     #
# See the LICENSE file or visit <http://www.gnu.org/licenses/>         #
########################################################################

SOURCE_PATH="/usr/lib/watchsys"

source $SOURCE_PATH/functions.sh

load_conf

showhelp()
{
	head
	echo 'Usage: watchsys [OPTION]'
	echo
	echo 'OPTIONS:'
	echo '-h | --help: Show this help screen'
	echo '-d | --start: Initialize a daemon for monitoring'
	echo '-s | --stop: Stop the daemon'
	echo '-t | --status: Show status of daemon and pid if currently running'
}

# Executed as a cleanup function when the daemon is stopped
on_daemon_exit()
{
	if [ -e /var/run/watchsys.pid ]; then
		rm -f /var/run/watchsys.pid
	fi
	
	exit 0
}

# Return the current process id of the daemon or 0 if not running
daemon_pid()
{
	if [ -e /var/run/watchsys.pid ]; then
		echo $(cat /var/run/watchsys.pid)
		
		return
	fi
	
	echo "0"
}

# Check if daemon us running.
# Outputs 1 if running 0 if not.
daemon_running()
{
	if [ -e /var/run/watchsys.pid ]; then
		running_pid=$(ps -A | grep watchsys | awk '{print $1}')
		
		if [ "$running_pid" != "" ]; then
			current_pid=$(daemon_pid)
			
			for pid_num in $running_pid; do
				if [ "$current_pid" = "$pid_num" ]; then
					echo "1"
					return
				fi
			done
		fi
	fi

	echo "0"
}

start_daemon()
{
	su_required
	
	if [ $(daemon_running) = "1" ]; then
		echo "watchsys daemon is already running..."
		exit 0
	fi
	
	echo "starting watchsys daemon..."
	
	nohup $0 -l > /dev/null 2>&1 &
	
	log_msg "daemon started"
}

stop_daemon()
{
	su_required
	
	if [ $(daemon_running) = "0" ]; then
		echo "watchsys daemon is not running..."
		exit 0
	fi
	
	echo "stopping watchsys daemon..."
	
	kill $(daemon_pid)
	
	while [ -e /var/run/watchsys.pid ]; do
		continue
	done
	
	log_msg "daemon stopped"
}

daemon_loop()
{
	su_required
	
	if [ $(daemon_running) = "1" ]; then
		exit 0
	fi
	
	if [ ! -d /var/cache/watchsys ]; then
		mkdir -p /var/cache/watchsys
		mkdir -p /var/cache/watchsys/timers
	fi
	
	echo "$$" > /var/run/watchsys.pid
	
	trap 'on_daemon_exit' INT
	trap 'on_daemon_exit' QUIT
	trap 'on_daemon_exit' TERM
	trap 'on_daemon_exit' EXIT
	
	# Sleep to allow any processes to properly start
	#sleep 60
	
	while true; do
		if [ $ALLOW_THREADING -eq 1 ]; then
			. $SOURCE_PATH/watch_cpu.sh &
			. $SOURCE_PATH/watch_mem.sh &
			. $SOURCE_PATH/watch_disk.sh &
			. $SOURCE_PATH/watch_proc.sh &
			. $SOURCE_PATH/watch_servers.sh &
		else
			monitor_cpu_usage
			monitor_memory_usage
			monitor_disk_usage
			monitor_services
			monitor_servers
		fi
		
		# Run monitors every 10 seconds
		sleep 10
	done
}

daemon_status()
{
	current_pid=$(daemon_pid)
	
	if [ $(daemon_running) = "1" ]; then
		echo "watchsys status: running with pid $current_pid"
	else
		echo "watchsys status: not running"
	fi
}

while [ $1 ]; do
	case $1 in
		'-h' | '--help' | '?' )
			showhelp
			exit
			;;
		'--start' | '-d' )
			start_daemon
			exit
			;;
		'--stop' | '-s' )
			stop_daemon
			exit
			;;
		'--status' | '-t' )
			daemon_status
			exit
			;;
		'--loop' | '-l' )
			# start daemon loop, used internally by --start | -s
			daemon_loop
			exit
			;;
		* )
			showhelp
			exit
			;;
	esac
	
	shift
done

showhelp

exit 0
